### PR TITLE
added FA2 and FA12 to Trustdart

### PR DESCRIPTION
**Pull request by Rakeally**

### CODE REVIEW

File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt [hunk 1]
```diff
@@ -31,4 +37,114 @@ class XTZ : Coin("XTZ", CoinType.TEZOS) {
...
+
+                val transactionData = TransactionOperationData.newBuilder()
+                    .setAmount((txData["transactionAmount"] as Int).toLong())
+                    .setDestination(txData["destination"] as String)
+                    .setParameters(parameters)
+                    .build()
+
+                val transaction = Operation.newBuilder()
+                    .setSource(txData["source"] as String)
+                    .setFee((txData["fee"] as Int).toLong())
```
Comment (jayluxferro) [inline android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt:81]: You can do this straight away.

```
txData["fee"]!!.toLong()
```

File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt [hunk 2]
```diff
@@ -31,4 +37,114 @@ class XTZ : Coin("XTZ", CoinType.TEZOS) {
...
+
+                val transactionData = TransactionOperationData.newBuilder()
+                    .setAmount(0)
+                    .setDestination(txData["destination"] as String)
+                    .setParameters(parameters)
+                    .build()
+
+                val transaction = Operation.newBuilder()
+                    .setSource(txData["source"] as String)
+                    .setFee((txData["fee"] as Int).toLong())
```
Comment (jayluxferro) [inline android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt:122]: `txData["fee"]!!.toLong()`

File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt [hunk 3]
```diff
@@ -31,4 +37,114 @@ class XTZ : Coin("XTZ", CoinType.TEZOS) {
...
+                    .setCounter((txData["counter"] as Int).toLong())
```
Comment (jayluxferro) [inline android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt:123]: `txData["counter"]!!.toLong()`

File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt [hunk 4]
```diff
@@ -31,4 +37,114 @@ class XTZ : Coin("XTZ", CoinType.TEZOS) {
...
+                    .setGasLimit((txData["gasLimit"] as Int).toLong())
```
Comment (jayluxferro) [inline android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt:124]: txData["gasLimit"]!!.toLong()

File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt [hunk 5]
```diff
@@ -31,4 +37,114 @@ class XTZ : Coin("XTZ", CoinType.TEZOS) {
...
+                    .setStorageLimit((txData["storageLimit"] as Int).toLong())
```
Comment (jayluxferro) [inline android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt:125]: txData["storageLimit"]!!.toLong()

**Review by jayluxferro — APPROVED** [review]

**Review by jayluxferro — CHANGES_REQUESTED** [review]
@Rakeally , you can cast direct too long. This helps prevent issues we had earlier with some of the tokens.

Example:

```
txData["counter"]!!.toLong()
```

**Review by baahkusi — CHANGES_REQUESTED** [review]
This PR seems to ignore native tezos transactions.

**Review by baahkusi — CHANGES_REQUESTED** [review]
It also doesn't take into account the case where user needs to do reveal

**Review by baahkusi — CHANGES_REQUESTED** [review]
It seems it's always doing a reveal. Reveal should only be done when specified.

File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt [hunk 6]
```diff
@@ -31,4 +40,159 @@ class XTZ : Coin("XTZ", CoinType.TEZOS) {
...
+                        .setSource(txData["source"] as String)
+                        .setFee(txData["fee"]!!.toLong())
+                        .setCounter(txData["counter"]!!.toLong())
+                        .setGasLimit(txData["gasLimit"]!!.toLong())
+                        .setStorageLimit(txData["storageLimit"]!!.toLong())
+                        .setKind(Operation.OperationKind.REVEAL)
+                        .setRevealOperationData(revealOperationData)
+                        .build()
+
+                    operationList.operationsList.add(reveal)
```
Comment (baahkusi) [inline android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt:111]: fee, counter, gasLimit and storageLimit are different for reveal so have their versions.

File: ios/Classes/coins/XTZ.swift [hunk 7]
```diff
@@ -27,5 +27,132 @@ class XTZ: Coin  {
...
+                        revealOperation.source = txData["source"] as! String;
+                        revealOperation.fee = txData["fee"] as! Int64
+                        revealOperation.counter = txData["counter"] as! Int64
+                        revealOperation.gasLimit = txData["gasLimit"] as! Int64
+                        revealOperation.storageLimit = txData["storageLimit"] as! Int64
+                        revealOperation.kind = .reveal
+                        revealOperation.revealOperationData = revealOperationData
+
+                    operationList.operations = [ transactionOperation, revealOperation ]
+
```
Comment (baahkusi) [inline ios/Classes/coins/XTZ.swift:136]: fee, counter, gasLimit and storageLimit are different for reveal so have their versions.

**Pull request merged**